lib/pg_party/model_decorator.rb

```diff
@@ -16,25 +18,27 @@ def partition_table_exists?
       connection.schema_cache.send(table_exists_method, target_table)
     end
 
-    def in_partition(table_name)
-      Class.new(__getobj__) do
-        self.table_name = table_name
-
-        # to avoid argument errors when calling model_name
-        def self.name
-          superclass.name
-        end
-
-        # when returning records from a query, Rails
-        # allocates objects first, then initializes
-        def self.allocate
-          superclass.allocate
-        end
-
-        # creating and persisting new records from a child partition
-        # will ultimately insert into the parent partition table
-        def self.new(*args, &blk)
-          superclass.new(*args, &blk)
+    def in_partition(child_table_name)
+      PgParty::Cache.fetch_model(table_name, child_table_name) do
```

Comment:
so this is storing a model definition in the mutex?

Reply:
`fetch_model` first attempts to fetch the existing value from the cache. If the value doesn't exist, it executes the block and caches the value.

It's very similar to `fetch` when using the Rails cache: http://api.rubyonrails.org/classes/ActiveSupport/Cache/Store.html#method-i-fetch